File: src/02.string/02.reverse/03.557.go
package reverse

func ReverseWords(s string) string {
	reverse := func(sArr []byte, start int, end int) {
		for i, j := start, end; i < j; i, j = i+1, j-1 {
			sArr[i], sArr[j] = sArr[j], sArr[i]
		}
	}

	n := len(s)
	sByte := []byte(s)


	startIndex := 0

	for i := 0; i < n; i++ {
		if s[i] == ' ' {
			reverse(sByte, startIndex, i-1)
			startIndex = i + 1
		} else if i == n-1 {
			reverse(sByte, startIndex, n-1)
		}
	}

	return string(sByte)
}

File: src/02.string/02.reverse/03.557.ts
export function reverseWords(s: string): string {
    const reverse = (s: string[], start: number, end: number) => {
        for (let i = start, j = end; i < j; i++, j--) {
            const temp = s[i];
            s[i] = s[j];
            s[j] = temp;
        }
    };

    const n = s.length;

    const sArr: string[] = [];

    for (let i = 0; i < n; i++) {
        sArr.push(s[i]);
    }

    let startIndex = 0;

    for (let i = 0; i < n; i++) {
        if (sArr[i] === " ") {
            reverse(sArr, startIndex, i - 1);
            startIndex = i + 1;
        } else if (i === n - 1) {
            reverse(sArr, startIndex, n - 1);
        }
    }

    s = "";

    for (let i = 0; i < n; i++) {
        s += sArr[i];
    }

    return s;
}
